File: lab/assets/js/external_image/eros.ts
import { getToken } from "../../../../shared/js/jwt";
import { ExternalImageProvider } from "./types";

export const EROS_BASE_URL = `${process.env.EUPHROSYNE_TOOLS_API_URL}/eros`;

interface ErosImage {
  filmnbr: string;
  worknbr: string;
  czone: string;
  aimfilm: string;
  technique: string;
  dtfilm: string;
  caimfilm: string | undefined;
  plfilm: string;
  opfilm: string;
  owner: string;
  stock: string;
  filmtype: string;
  zone: string;
  Xsize: string;
  Ysize: string;
  bands: string;
}

interface ErosObject {
  title: string;
  local: string;
  owner: string;
  worknbr: string;
  collection: string;
  dtfrom: string;
  dtto: string;
  appel: string;
  support: string;
  technique: string;
  height: string;
  width: string;
  workgroup: string;
  srmf: string;
  categ: string;
  images: ErosImage[] | null;
}

function constructFromPath(path: string, token?: string) {
  const [erosId, imageId] = path.split("/");
  let imageCategory: string;
  if (erosId.startsWith("C2RMF")) {
    imageCategory = `pyr-${erosId.substring(0, 6)}`;
  } else if (erosId.startsWith("F")) {
    imageCategory = `pyr-${erosId.substring(0, 2)}`;
  } else {
    imageCategory = `pyr-FZ`;
  }
  let url = `${EROS_BASE_URL}/iiif/${imageCategory}/${erosId}/${imageId}.tif/full/500,/0/default.jpg`;
  if (token) {
    url = `${url}?token=${token}`;
  }
  return url;
}

async function getImagesURLForObject(providerObjectId: string) {
  const objectDetailsURL = `${EROS_BASE_URL}/rails/oeuvres/${providerObjectId}.json`;
  let fetchFailed = false,
    objectResponse: Response | undefined = undefined;
  const token = await getToken();
  try {
    objectResponse = await fetch(`${objectDetailsURL}?token=${token}`);
  } catch (error) {
    console.error(error);
    fetchFailed = true;
  }
  if (fetchFailed || (objectResponse && !objectResponse.ok)) {
    console.warn(
      `Failed to fetch object details with id ${providerObjectId}.\n
        URL: ${objectDetailsURL}\n`,
    );
    if (objectResponse && !objectResponse.ok) {
      console.warn(`
        Status: ${objectResponse.statusText}\n
        Content: ${await objectResponse.text()}`);
    }
    return [];
  }
  if (objectResponse) {
    const objectDetails = (await objectResponse.json()) as ErosObject;
    if (!objectDetails.images || objectDetails.images.length === 0) {
      console.warn(`No images found for object with id ${providerObjectId}`);
      return [];
    }
    return objectDetails.images.map((image) =>
      constructFromPath(`${providerObjectId}/${image.filmnbr}`, token),
    );
  }
  return [];
}

export const erosImageService: ExternalImageProvider = {
  getImagesURL: (providerObjectId: string) =>
    getImagesURLForObject(providerObjectId),
  constructFromPath: (path: string, token?: string) =>
    constructFromPath(path, token),
};

File: lab/objects/providers/eros.py
import os
import typing
from functools import lru_cache

import requests
from django.conf import settings

from lab.thesauri.models import Era

from ..models import ObjectGroup
from .base import ObjectProvider
from .registry import registry

EROS_BASE_URL = "http://eros.c2rmf.fr"


class ErosImage(typing.TypedDict):
    """EROS image data structure."""

    filmnbr: str
    worknbr: str
    czone: str
    aimfilm: str
    technique: str
    dtfilm: str
    plfilm: str
    opfilm: str
    owner: str
    stock: str
    filmtype: str
    zone: str
    Xsize: str
    Ysize: str
    bands: str


class ErosData(typing.TypedDict):
    """EROS object data structure."""

    title: str
    local: str
    owner: str
    worknbr: str
    collection: str
    dtfrom: str
    dtto: str
    appel: str
    support: str
    technique: str
    height: str
    width: str
    workgroup: str
    srmf: str
    categ: str
    inv: str
    period: str
    images: typing.NotRequired[list[ErosImage]]


class ErosProvider(ObjectProvider):
    """C2RMF EROS provider implementation."""

    @lru_cache
    def _fetch_raw_data(self, eros_id: str) -> ErosData | None:
        """Fetch raw object data from EROS."""
        token = os.environ["EROS_HTTP_TOKEN"]
        try:
            request = requests.get(
                f"{EROS_BASE_URL}/rails/oeuvres/{eros_id}.json?token={token}",
                timeout=5,
            )
            request.raise_for_status()
        except (requests.HTTPError, requests.ConnectionError) as error:
            self._handle_http_error(error, eros_id)

        if request.status_code == 200:
            return request.json()
        return None

    def fetch_partial_data(self, object_id: str):
        """Fetch object data with minimum information to save it to DB."""
        data = self._fetch_raw_data(object_id)
        if data:
            return {"label": data["title"]}
        return None

    def fetch_full_object(
        self, object_id: str, object_group: ObjectGroup | None = None
    ) -> ObjectGroup | None:
        """Fetch object data with full information to display it."""
        updated_og = object_group or ObjectGroup()
        data = self._fetch_raw_data(object_id)
        if not data:
            return object_group

        updated_og.object_count = 1
        updated_og.label = data["title"]
        if data.get("dtfrom") or data.get("period"):
            dating_label = data.get("dtfrom") or data["period"]
            updated_og.dating_era = Era(label=dating_label)
        updated_og.collection = data.get("collection") or ""
        updated_og.inventory = data.get("inv") or ""
        updated_og.materials = (data.get("support") or "").split(" / ")
        return updated_og

    def fetch_object_image_urls(self, object_id: str) -> list[str]:
        """Fetch list of image URLs for the given object ID."""
        data = self._fetch_raw_data(object_id)
        if not data or "images" not in data:
            return []
        return [
            self.construct_image_url(f"{object_id}/{image['filmnbr']}")
            for image in data["images"]
        ]

    def construct_image_url(self, path: str) -> str:
        """Construct image URL from C2RMF path."""
        # pylint: disable=import-outside-toplevel
        from euphro_auth.jwt.tokens import EuphroToolsAPIToken

        eros_id, image_id = path.split("/")
        if eros_id.startswith("C2RMF"):
            image_category = f"pyr-{eros_id[:6]}"
        elif eros_id.startswith("F"):
            image_category = f"pyr-{eros_id[:2]}"
        else:
            image_category = "pyr-FZ"

        eros_base_url = (
            settings.EROS_BASE_IMAGE_URL or f"{settings.EUPHROSYNE_TOOLS_API_URL}/eros"
        )

        url = f"{eros_base_url}/iiif/{image_category}/{eros_id}/{image_id}.tif/full/500,/0/default.jpg"  # pylint: disable=line-too-long

        # Add token to the URL if using EROS direct URL. Else we use the EuphroTools API
        # proxy which includes the token in the request headers.
        if settings.EROS_BASE_IMAGE_URL:
            token = os.environ["EROS_HTTP_TOKEN"]
        else:
            token = EuphroToolsAPIToken.for_euphrosyne().access_token
        return f"{url}?token={token}"


# Register the EROS provider
registry.register("eros", ErosProvider)
